Audit broken in 0.7.1

Since the upgrade from 0.7.0 to 0.7.1 we ran into the following issue:

    client$ VAULT_TOKEN=<redacted> vault token-lookup
    error looking up token: Error making API request.
    
    URL: GET https://vault.example.com:8200/v1/auth/token/lookup-self
    Code: 500. Errors:
    
    * internal error

The Vault daemon reports:

    May 08 15:26:30 vault-101.example.com vault[53455]: 2017/05/08 15:26:30.836345 [ERROR] audit: panic during logging: request_path=auth/token/lookup-self error="reflect: reflect.Value.Set using value obtained using unexported field"
    May 08 15:26:30 vault-101.example.com vault[53455]: 2017/05/08 15:26:30.836397 [ERROR] core: failed to audit response: request_path=auth/token/lookup-self error=panic generating audit log

Using the root token still works:

    server# vault token-lookup | grep '^path'
    path            	auth/token/root

Enabled audit backend:

    server# vault audit-list
    Path   Type  Description  Replication Behavior  Options
file/  file               replicated            file_path=/var/log/vault/audit.log

If we disable audit logging, the 500 for the client disappears:

    server# vault audit-disable file
    Successfully disabled audit backend 'file' if it was enabled
    client$ VAULT_TOKEN=<redacted> vault token-lookup | grep '^path'
    path            	auth/token/create

Also when we enable the syslog backend, the client receives a 500 and the same error is reported by the Vault daemon.

--- a/audit/hashstructure.go
+++ b/audit/hashstructure.go
@@ -1,8 +1,10 @@
 package audit
 
 import (
+	"errors"
 	"reflect"
 	"strings"
+	"time"
 
 	"github.com/hashicorp/vault/helper/salt"
 	"github.com/hashicorp/vault/helper/wrapping"
@@ -141,6 +143,12 @@ type hashWalker struct {
 	unknownKeys []string
 }
 
+// hashTimeType stores a pre-computed reflect.Type for a time.Time so
+// we can quickly compare in hashWalker.Struct. We create an empty/invalid
+// time.Time{} so we don't need to incur any additional startup cost vs.
+// Now() or Unix().
+var hashTimeType = reflect.TypeOf(time.Time{})
+
 func (w *hashWalker) Enter(loc reflectwalk.Location) error {
 	w.loc = loc
 	return nil
@@ -185,6 +193,41 @@ func (w *hashWalker) Slice(s reflect.Value) error {
 func (w *hashWalker) SliceElem(i int, elem reflect.Value) error {
 	w.csKey = append(w.csKey, reflect.ValueOf(i))
 	w.sliceIndex = i
+	return nil
+}
+
+func (w *hashWalker) Struct(v reflect.Value) error {
+	// We are looking for time values. If it isn't one, ignore it.
+	if v.Type() != hashTimeType {
+		return nil
+	}
+
+	// If we aren't in a map value, return an error to prevent a panic
+	if v.Interface() != w.lastValue.Interface() {
+		return errors.New("time.Time value in a non map key cannot be hashed for audits")
+	}
+
+	// Override location to be a MapValue. loc is set to None since we
+	// already "entered" the struct. We could do better here by keeping
+	// a stack of locations and checking the last entry.
+	w.loc = reflectwalk.MapValue
+
+	// Create a string value of the time. IMPORTANT: this must never change
+	// across Vault versions or the hash value of equivalent time.Time will
+	// change.
+	strVal := v.Interface().(time.Time).UTC().Format(time.RFC3339Nano)
+
+	// Walk it as if it were a primitive value with the string value.
+	// This will replace the currenty map value (which is a time.Time).
+	if err := w.Primitive(reflect.ValueOf(strVal)); err != nil {
+		return err
+	}
+
+	// Skip this entry so that we don't walk the struct.
+	return reflectwalk.SkipEntry
+}
+
+func (w *hashWalker) StructField(reflect.StructField, reflect.Value) error {
 	return nil
 }
 

--- a/audit/hashstructure_test.go
+++ b/audit/hashstructure_test.go
@@ -140,6 +140,10 @@ func TestHash(t *testing.T) {
 			&logical.Response{
 				Data: map[string]interface{}{
 					"foo": "bar",
+
+					// Responses can contain time values, so test that with
+					// a known fixed value.
+					"bar": time.Unix(1494264707, 0),
 				},
 				WrapInfo: &wrapping.ResponseWrapInfo{
 					TTL:             60,
@@ -151,6 +155,7 @@ func TestHash(t *testing.T) {
 			&logical.Response{
 				Data: map[string]interface{}{
 					"foo": "hmac-sha256:f9320baf0249169e73850cd6156ded0106e2bb6ad8cab01b7bbbebe6d1065317",
+					"bar": "hmac-sha256:b09b815a7d1c3bbcf702f9c9a50ef6408d0935bea0154383a128ca8743eb06fc",
 				},
 				WrapInfo: &wrapping.ResponseWrapInfo{
 					TTL:             60,

